### ISSUE TITLE
Question: Snowflake performance in SQLPro Studio vs Web UI

### ISSUE THREAD

**Issue by jim256** [hankinsoft/SQLPro#428]
Should I expect to see any differences in execution time between one and the other? I'm not quite sure how the web UI does it and if it's actually returning all results, or just some, or what.

I have seen a lot of times that the web UI will be significantly faster. As an example, I just ran a pretty simply SELECT * query with 2 WHERE conditions. SQLPro Studio took 1 min 49 seconds to return 314,912 records. The web UI returned the same number of records with a time of 131ms. I've noticed this kind of thing a lot, but don't understand the underlying concepts well enough to know if it's even a problem.

When the web UI runs it so quickly, I sometimes wonder if it's because it hasn't loaded all the results, or because of caching.

To test this, I scrolled all the way to the bottom of the web UI results. It seems to have loaded them all. I also tried exporting the data to CSV. This took about 27 seconds, generating a 63MB CSV file that downloaded in about 3 seconds.

I tried repeating the same query again in both the web UI and in the app, and both took the same amount of time the second time around. I tried a new query, this time running it in the web UI first (in case it was somehow caching the results from the query run from the app, then using that to load the results in the web interface), and the performance comparison didn't change much (46k records, web UI 1.96s, SQLPro Studio 14.139 seconds).

I've also noticed that the default SELECT * FROM [TABLE_NAME] LIMIT 1000; query often takes about 10 seconds to run, which always surprises me.
I've attached the log file of running one of these queries in case it's of any help. 3.478 seconds in SQLPro Studio and 106ms in the web UI.
[com.hankinsoft.osx.sqlprostudio 2020-05-05--17-15-43-685.log](https://github.com/hankinsoft/SQLPro/files/4582571/com.hankinsoft.osx.sqlprostudio.2020-05-05--17-15-43-685.log)

Thoughts? Is this expected?

My connection at the moment has a ping of 12ms and download of 125Mbps, which is relatively typical.

SQLPro Studio 2020.39 (Build 110666.5)
macOS 10.15.4 (19E287)
Snowflake 4.13.2

**Comment by jim256** [comment]
As an aside, I'd love to get your thoughts on all the SSH issues present in my log. All of my SSH connections seem to work fine, but I get the impression there are issues given how many SSH problems it mentions. I just don't really know what those are / if they are anything to care about.

**Comment by hankinsoft** [comment]
What do you mean by SSH issues? Could you post a small example?

**Comment by jim256** [comment]
@hankinsoft just wanted to bump this guy up. The speed difference is kind of killer sometimes.

It's been kind of pushing me to different clients (typically the web console, which is a pretty sorry experience). When I'm just double-clicking on a table in the explorer to quickly profile some data, for example,  and each one takes ~13 seconds to return the standard 1000 rows, the 13 seconds * number of tables I'm looking at adds up to a lot more time than I actually need to look at the data.

**Comment by hankinsoft** [comment]
I haven't had a chance to look into this one yet. I'm hoping to dig into it within the next couple of weeks.

**Comment by hankinsoft** [comment]
If you get a chance, please give the latest version a chance and let me know if you find its running any quicker.

**Comment by jim256** [comment]
@hankinsoft, it could be slightly faster, though if so, it would be a very minor way.

One good way to test it is just running "SELECT 1". For me, that will take anywhere from 2-8 seconds. I'm not sure what determines that. I wondered if it was the warehouse being suspended and there an auto-resume delay, but I'm always testing on an already-running warehouse.

In contrast, running "SELECT 1" on my Postgres or MySQL connections always seems to complete in 1-2 seconds.

One piece of this I've noticed (which should maybe be a separate issue) is that the query doesn't seem to get "sent" for about 1 second. If I hit execute, and then immediately thereafter change the query to be invalid, the execution will fail. Or if I hit execute, and then hurry and paste in an additional column name in the SELECT, that additional column will come back in the results.

**Comment by hankinsoft** [comment]
FYI I've been wresting with this one and unfortunately I've been unable to get a performance improvement. From what I can tell, all communication is done over HTTPS and using base64 encrypted data. It seems most of the time spent is within the [libsnowflakeclient](https://github.com/snowflakedb/libsnowflakeclient) library reading/writing via CURL and/or waiting to send/receive data.

<img width="1224" alt="Screen Shot 2020-06-30 at 2 16 40 PM" src="https://user-images.githubusercontent.com/3160528/86156459-5df30300-badc-11ea-8e50-06aa6c724d45.png">

Snowflake does provide a few other ways to connect, however none are currently compatible with SQLPro (think JDBC).

I'll need to take some time to think this over a bit more - but as it stands, I don't currently have a way to improve the snowflake performance.

**Comment by jim256** [comment]
@hankinsoft, understood. I appreciate your digging into it so thoroughly. Hopefully something will come to light at some point down the road.

What were your thoughts on this?:
> One piece of this I've noticed (which should maybe be a separate issue) is that the query doesn't seem to get "sent" for about 1 second. If I hit execute, and then immediately thereafter change the query to be invalid, the execution will fail. Or if I hit execute, and then hurry and paste in an additional column name in the SELECT, that additional column will come back in the results.

I was thinking:
a) this could be part of the performance issue, since the query execution therefore must not even begin for at least ~1 second.
b) I'd personally prefer to see this behavior change. It's not very intuitive to me to have to remember to wait a bit after hitting execute before I can start modifying the text in the editor. I'm often testing different things when running queries, so I'll hit execute, and then go to start making changes to prepare for the next execution, only to have those changes picked up in the current execution.

**Comment by jim256** [comment]
Just as another note here (maybe this will be helpful down the road), I just ran a simple:
```
SELECT
	LEN(mt.COL_NAME)
,	mt.*
	
FROM
	my_table mt
```

Took 9 min and 3 seconds in SQLPro Studio on Snowflake
It took 31 seconds on the source Postgres database (same table structure and data).

My computer's fan was running at max speed and sounded like it was prepping for takeoff during the Snowflake query.

The Snowflake web UI claimed the same query took 196ms.

TablePlus performed similarly to SQLPro Studio.
DBeaver (which I think uses JDBC) took 1min 50 seconds.

Table has 560k rows in it. The max length of that field was 300 chars. As you can imagine, waiting 9.5min would be brutal.

**Comment by hankinsoft** [comment]
> What were your thoughts on this?:
>> One piece of this I've noticed...

Sorry, I should have commented on this when I made the other note as I did a bit of research into it.

Yes, I can improve this a little bit and fix the 'changing text after hitting execute' issue. What is happening here, is that when you hit execute, a new background thread gets started which does all of the heavy lifting. A few things happen before the query actually gets executed, such as:

1. Detect if the user has a selection range, and if so use that as the query.
2. If not, check if the user has 'execute query in range' turned on and calculate the proper query to run.
3. Save the query to query history.
4. Check if the connection is still active by pinging it (it could have closed due to timeout or some other thing). If its not active, then re-connect.
5. Send the query to the server.

1-4 all happen the same for all database types, then the server specific implementation kicks off at step 5.

You can change the text for two reasons:
1. Previously the new thread was set as default priority. I'm going to raise this priority.
2. The thread is currently sent a pointer to the query string. Meaning if you change it before it gets to a certain state, it was using the updated value. I've got a few things I can do here, with the easiest being 'create a copy of the query'. That could cause performance issues if the user has multi-mb file open.

With all this being said --- Yes, I can/will fix that portion of the code and it might give a small performance increase (probably less than 1 second). It will also fix typing after you hit execute. Not for the next build (which I'm submitting now), but the build after).

**Comment by hankinsoft** [comment]
> Not for the next build (which I'm submitting now), but the build after).

Scratch that - I have made some improvements around this for the next build 2020.61 which is being submitted to the App Store now.

**Comment by hankinsoft** [comment]
2020.61 is now available which adds the slight performance improvement 'at query startup'. For the most part you probably won't see a huge difference (~.50 - 1 second). If you get the chance, please give it a test and confirm you can no longer edit the text before the 'query actually executes'.

**Comment by jim256** [comment]
@hankinsoft, seems to have done the trick in terms of post-execute changes taking effect in execution!

**Comment by hankinsoft** [comment]
If you get the chance, can you try duplicating your snowlflake connector and adding the text 'JDBC' into the alias field? After connecting, you should see the term 'JDBC' before the server version in the database tree. Is the query performance any better on this? (NOTE: CANCELLING A QUERY MIGHT NOT WORK 100% - you might need to cancel long running queries via the webui to prevent additional charges as the JDBC support is beta).

**Comment by jim256** [comment]
Tried connecting, but got this:
<img width="408" alt="image" src="https://user-images.githubusercontent.com/15026225/96776164-dfc9ad80-13a5-11eb-8eae-3368ffbe5a3e.png">

[com.hankinsoft.osx.sqlprostudio 2020-10-21--20-00-22-533.log](https://github.com/hankinsoft/SQLPro/files/5418445/com.hankinsoft.osx.sqlprostudio.2020-10-21--20-00-22-533.log)

macOS 10.15.7 (19H2)
Snowflake SQLPro Studio 2020.88 (Build 110729.5)
Snowflake 4.34.3

**Comment by jim256** [comment]
FWIW, just tried this again on different versions of everything and still get the same thing.

macOS 11.0.1 (20B29)
SQLPro Studio 2020.97 (Build 110745.5)
Snowflake 4.39.2

**Comment by jim256** [comment]
Failure log when trying from SQLPro for Snowflake test build:
[com.hankinsoft.macos.sqlpro-for-snowflake 2020-11-19--23-32-09-087.log](https://github.com/hankinsoft/SQLPro/files/5570309/com.hankinsoft.macos.sqlpro-for-snowflake.2020-11-19--23-32-09-087.log)


SQLPro for Snowflake 2020.101 (Build 7)
macOS 11.0.1 (20B29)
Snowflake 4.39.5

**Comment by jim256** [comment]
FWIW, that _credits_per_query query from #408 can be good testing for the performance as well.

Once that table is made, if I do a simple SELECT * LIMIT 5000, it takes about 13 seconds in SQLPro Studio, while only taking 2.88 seconds in the Snowflake web console. I thought this was interesting as it was strictly data retrieval.

**Comment by jim256** [comment]
This one is still a killer for me. Even just expanding a Snowflake database node to see what schemas are inside it will usually take around 8 seconds when it takes less than 1 in the web UI. Returning query results is still much, much slower than the web UI as well.

Having this and temp table autocomplete would really speed up the workflow of me and my team pretty significantly.

**Comment by richardo2008** [comment]
This would be awesome! It would really speed up things for me in Snowflake as well.

**Comment by hankinsoft** [comment]
Can you try this build:
https://d3fwkemdw8spx3.cloudfront.net/temp/SQLProSnowflake-S1.zip

Using the following query as a test:

```
use schema public;
CREATE OR REPLACE TEMP TABLE _credits_per_query AS

	SELECT
		mh.NAME WAREHOUSE
	,	DATE_TRUNC('MONTH', mh.START_TIME)::DATE USAGE_MONTH
	,	mh.SERVICE_TYPE
	,	mh.START_TIME
	,	mh.END_TIME
	,	mh.CREDITS_USED
	,	qh.QUERY_ID
	,	qh.QUERY_TEXT
	,	qh.QUERY_TYPE
	,	qh.USER_NAME
	,	qh.ROLE_NAME
	,	qh.START_TIME QUERY_START_TIME
	,	qh.END_TIME QUERY_END_TIME
	,	qh.EXECUTION_TIME
	,	(qh.EXECUTION_TIME / SUM(qh.EXECUTION_TIME) OVER (PARTITION BY mh.NAME, mh.START_TIME, mh.END_TIME)) * 100 PCT_CREDIT_USED_BY_QUERY
	,	((PCT_CREDIT_USED_BY_QUERY / 100) * mh.CREDITS_USED) CREDIT_USED_BY_QUERY

	FROM
		SNOWFLAKE.ACCOUNT_USAGE.METERING_HISTORY mh
		JOIN SNOWFLAKE.ACCOUNT_USAGE.QUERY_HISTORY qh ON mh.ENTITY_ID = qh.WAREHOUSE_ID
			AND qh.START_TIME BETWEEN mh.START_TIME AND mh.END_TIME

	WHERE
		TRUE
	AND	qh.WAREHOUSE_SIZE IS NOT NULL -- Eliminate queries served by Cloud Services layer
	AND mh.START_TIME > DATE_TRUNC('MONTH', ADD_MONTHS(CURRENT_DATE(), -6)) -- Edit as per analysis period
--	AND	qh.EXECUTION_STATUS = 'SUCCESS' -- Disabled since failed execution will also consume compute
;


-- Once the temp table above has been completed, the below 20,000 now comes back in 3.5s
-- vs over a minute before.
SELECT
	*
FROM
	_credits_per_query AS a JOIN _credits_per_query AS b ON 1 = 1 LIMIT 20000
```

The last select now comes back at 3.5s vs over a minute beforehand.

**Comment by jim256** [comment]
Sounds awesome! I just tried it, but it's asking for a license key. Since I'm on the trial of SQLPro Studio, I don't have one to use.

I still did the create (but limited it to 3 months of data), then tried the select, but the select has been running for 27 min and hasn't retrieved any records yet. Double-clicking tables for the standard select seems to work (though I obviously can't see the results).

**Comment by hankinsoft** [comment]
Can you try https://d3fwkemdw8spx3.cloudfront.net/temp/SQLProSnowflake-S2.zip?
Please let (a shorter query if possible) complete, and provide the logs (they can be emailed). It should show me where most of the time is being spent.

**Comment by hankinsoft** [comment]
Hi,
It looks like that statement might be quite large server side. (The too much memory comes from the server I believe). I would recommend maybe a smaller query, or one of your own system that you know of that you have seen run slow in SQLPro but faster in the UI.

**Comment by jim256** [comment]
Here's one run with the new build with a 1-month lookback period
<img width="1130" alt="image" src="https://user-images.githubusercontent.com/15026225/113199696-a898db80-9224-11eb-8aaa-13fba25ecccc.png">


[com.hankinsoft.macos.sqlpro-for-snowflake 2021-03-31--17-06-46-637.log](https://github.com/hankinsoft/SQLPro/files/6238989/com.hankinsoft.macos.sqlpro-for-snowflake.2021-03-31--17-06-46-637.log)

**Comment by jim256** [comment]
Just ran again, but without the self join, so it'd be much smaller / faster. A couple things:

It presented the following message even as (as you can see at the bottom) it was still "running" the query, which seemed odd. About 10 seconds after I took this screenshot, the app switched to the results grid, the dock icon bounced, and the bottom bar said "Finished".
<img width="1145" alt="image" src="https://user-images.githubusercontent.com/15026225/113200805-07128980-9226-11eb-8d57-83d2ead8cc35.png">

My CPU (2019 16" i9 MBP) ranged from 65% - 132%, usually at about 85%, with the fan running the whole time. Memory usage steadily climbed until it reached 27GB.

[com.hankinsoft.macos.sqlpro-for-snowflake 2021-03-31--19-08-11-623.log](https://github.com/hankinsoft/SQLPro/files/6239059/com.hankinsoft.macos.sqlpro-for-snowflake.2021-03-31--19-08-11-623.log)

**Comment by hankinsoft** [comment]
Before I got further, how is the actual performance - if you run that same statement on SQLPro studio, does it take longer?

**Comment by jim256** [comment]
I was testing this right after that one. It seemed that querying an existing table was about the same between the two, whereas querying a temp table was much faster in the new build.
I'm going to verify and get stats and logs for you later (probably tomorrow).

**Comment by hankinsoft** [comment]
Ok - FYI from my tests I believe the slowdown is specifically with tables
that include timestamp columns.

**Comment by jim256** [comment]
Ah, that's helpful. Really, thank you so much for digging in deeply here.

I seem to be seeing about the same thing. If I select just IDs (integers), the new build is slightly faster.
If I select a column with datetimes, the new build is MUCH faster! Seems to be by a factor of 5 - 12 (roughly speaking, obviously). This is huge!

Just for reference, on a 1M-row query of a datetime column only:
SQLPro Studio: 3 min, 7 seconds
New build of SQLPro for Snowflake : 18.7 seconds
Snowflake web console: 67ms

I'm not 100% sure if the Snowflake web console is a fair comparison. I say this because, if you scroll quickly in the results pane, it'll say "Loading" for up to 2 seconds on the data. So I'm not sure if it's actually retrieved them all or if it's paging. The real comparison would be to download DBeaver on this laptop and disable paging with that client, but I don't have time this morning.

Again, this is such promising progress so far, and I really appreciate it! This will semi-significantly speed up the team's work every single day, which will be fantastic.

**Comment by hankinsoft** [comment]
> The real comparison would be to download DBeaver on this laptop and disable paging with that client, but I don't have time this morning.

When you get the chance, please do let me know.

I've got a bit more work to do with this, but I plan on getting it into the main SQLPro Studio build next week.

**Comment by jim256** [comment]
Holy smokes I think you've almost done it.

On a table with datetimes, SQLPro for Snowflake was 33 seconds for 100k rows. DBeaver was 2:18. Snowflake web console claimed 997ms, but when I try to export the data to CSV (to make sure it's actually been transferred to me), it took 33 seconds to export it, so I think it is paging.

On a table without datetimes, DBeaver and SQLPro for Snowflake were both 22 seconds.

If I do a SELECT with a LIMIT 100, the Snowflake web console is 328ms, and I think that's legitimate, because it seems unlikely that the paging would be smaller than that, and I can scroll anywhere in the result set without getting "Loading".
SQLPro for Snowflake is 1.7 seconds.
DBeaver is about 400ms.

I know that you've mentioned some time is consumed on saving the query to history and a few other steps. FWIW, it looks like Snowflake's web console and DBeaver both also keep a log with things like the query text, duration, etc.

This is honestly huge.

**Comment by hankinsoft** [comment]
The latest SQLPro Studio build includes this, plus a few minor additional performance improvements. If you get the chance, please give it a try. This will not affect anything to down with the setup/teardown performance of running a query.

**Comment by jim256** [comment]
Awesome! This is great.

I think there are two impactful items remaining here:
1. Web console vs SQLPro Studio with the default LIMIT 100 (in the case of the browser UI): Web UI is about 80ms while SQLPro Studio is about 3.1 seconds. SQLPro Studio is super fast with Postgres and MySQL.
2. Expanding the nodes on the left. Opening a schema to see the tables inside takes less than 1 second in the web UI and about 4 seconds. SQLPro Studio is super fast with Postgres and MySQL.

**Comment by hankinsoft** [comment]
Unfortunately, I don't currently have any ideas on how to improve the performance for these two issues. If you take a look at the following screenshot:

<img width="1826" alt="Screen Shot 2021-04-07 at 4 00 33 PM" src="https://user-images.githubusercontent.com/3160528/113920053-aac9df80-97ba-11eb-80ac-61bf72bdb81a.png">

It shows roughly 90% of the time during an 'execute query' is being taken up in the actual snowflake database library calling curl commands (reaching out to snowflake to connect & run the query).

I was able to improve the timestamp performance as those were the 'on client machine' processing of data. I wrote my own processing which was more performant than the processing included in snowflakes library. These curl calls are the actual network calls. Currently I can't think of anyway I can improve the performance around these portions.

**Comment by jim256** [comment]
Wow, that's awesome that your own processing outperformed Snowflake's.

I can see being stuck on query speed, and I can't get any better performance from DBeaver or TablePlus on query execution. Maybe Snowflake deliberately doesn't process queries from other clients as quickly or something. A shame since Snowflake can be really fast and SQLPro Studio can be really fast, but maybe the combo just can't be done if Snowflake doesn't expose endpoints / resources for it.

Think it's worth raising an issue with Snowflake's repos regarding a) your faster processing of timestamp columns or b) the fact that the same query runs so much faster in their web console vs anything else? Maybe there's some optimization they'd be willing to do on their end that wouldn't require any work from your end?

When it comes to expanding the nodes on the left, though, DBeaver is significantly faster. I'd imagine that means there's a way. I'll frequently try to guess table names or rely on autocomplete rather than wait on expanding the nodes if I'm trying to look through a couple quickly to double-click on a tables to explore them. Not sure how helpful that is. For what it's worth, DBeaver has a community edition (free, that's what I'm using) that's super-quick to set up and test.

**Comment by hankinsoft** [comment]
I've added some additional logging in the latest version. Could you close all SQLPro connections/tabs, then re-launch and expand the schemas and table nodes, then provide me with the log files?

**Comment by jim256** [comment]
You got it. Attached.
[com.hankinsoft.osx.sqlprostudio 2021-04-09--22-52-58-954.log](https://github.com/hankinsoft/SQLPro/files/6288830/com.hankinsoft.osx.sqlprostudio.2021-04-09--22-52-58-954.log)

**Comment by hankinsoft** [comment]
A new build is available via the Mac app store. Can you follow the same steps again and provide the updated log file?

**Comment by jim256** [comment]
Here's the latest log.
[com.hankinsoft.osx.sqlprostudio 2021-04-19--16-28-44-814.log](https://github.com/hankinsoft/SQLPro/files/6337650/com.hankinsoft.osx.sqlprostudio.2021-04-19--16-28-44-814.log)

Seriously, though, the improved query speed is so so nice. It's absolutely awesome to get results back in ~3 seconds rather than ~9 when exploring tables via double-clicking. Thanks again.

**Comment by hankinsoft** [comment]
Are you seeing better performance for the database tree expanding as of the latest build? I believe the change I made which improved the search performance would also improve the database tree performance. If you are NOT seeing better performance could you yet again post a clean log if/when you get the chance?

**Comment by jim256** [comment]
Seems to be inconsistent. Was about 3 seconds for a schema that's about .5 in the browser, but then it was about .5 next time after disconnecting/reconnecting for the same schema. That schema has fewer than 300 tables, and Snowflakes explorer only loads 100 by default each time. That said, I think it's still quite a bit faster (.5 * 3 < 3, and I'd imagine there was some repeated overhead there).
[com.hankinsoft.osx.sqlprostudio 2021-05-27--18-37-08-689.log](https://github.com/hankinsoft/SQLPro/files/6556074/com.hankinsoft.osx.sqlprostudio.2021-05-27--18-37-08-689.log)


Here's a clean log file for just opening the one that took about ~3 seconds total (manually timed on my phone). Took ~.75 seconds in the browser. Had no other connections open. Had no tabs open. Had closed and restarted the app, freshly opened just this connection, drilled into the DB, deleted the log file, then drilled into the schema.

**Comment by hankinsoft** [comment]
The latest build (released yesterday), I implemented a new feature in the database tree. Scheme and tree node expands should be near instantaneous. Let me know what you think.

**Comment by jim256** [comment]
This is AWESOME. Holy smokes it's fast. Thanks a lot.
